Mathlib/GroupTheory/CommutingProbability.lean

```diff
@@ -132,3 +136,90 @@ theorem inv_card_commutator_le_commProb : (↑(Nat.card (commutator G)))⁻¹ 
     (le_trans (ge_of_eq (commProb_eq_one_iff.mpr (Abelianization.commGroup G).mul_comm))
       (commutator G).commProb_quotient_le)
 #align inv_card_commutator_le_comm_prob inv_card_commutator_le_commProb
+
+-- Construction of group with commuting probability 1/n
+namespace DihedralGroup
+
+lemma commProb_odd {n : ℕ} (hn : Odd n) :
+    commProb (DihedralGroup n) = (n + 3) / (4 * n) := by
+  rw [commProb_def', DihedralGroup.card_conjClasses_odd hn, nat_card]
+  qify [show 2 ∣ n + 3 by rw [Nat.dvd_iff_mod_eq_zero, Nat.add_mod, Nat.odd_iff.mp hn]; rfl]
+  rw [div_div, ← mul_assoc]
+  congr
+
+private lemma div_two_lt {n : ℕ} (h0 : n ≠ 0) : n / 2 < n :=
+  Nat.div_lt_self (Nat.pos_of_ne_zero h0) (lt_add_one 1)
+
+private lemma div_four_lt : {n : ℕ} → (h0 : n ≠ 0) → (h1 : n ≠ 1) → n / 4 + 1 < n
+  | 0 | 1 | 2 | 3 => by decide
+  | n + 4 => by intros; linarith [n.add_div_right four_pos, n.div_le_self 4]
+
+/-- A list of Dihedral groups whose product will have commuting probability `1 / n`. -/
+def reciprocalFactors (n : ℕ) : List ℕ :=
+  if h0 : n = 0 then [0]
+  else if h1 : n = 1 then []
+  else if Even n then
+    have := div_two_lt h0
+    3 :: reciprocalFactors (n / 2)
+  else
+    have := div_four_lt h0 h1
+    n % 4 * n :: reciprocalFactors (n / 4 + 1)
+
+@[simp] lemma reciprocalFactors_zero : reciprocalFactors 0 = [0] := rfl
+
+@[simp] lemma reciprocalFactors_one : reciprocalFactors 1 = [] := rfl
+
+@[nolint unusedHavesSuffices] lemma reciprocalFactors_even {n : ℕ} (h0 : n ≠ 0) (h2 : Even n) :
+    reciprocalFactors n = 3 :: reciprocalFactors (n / 2) := by
+  have h1 : n ≠ 1
+  · rintro rfl
+    norm_num at h2
+  conv_lhs => unfold reciprocalFactors
+  rw [dif_neg h0, dif_neg h1, if_pos h2]
+
+@[nolint unusedHavesSuffices] lemma reciprocalFactors_odd {n : ℕ} (h1 : n ≠ 1) (h2 : Odd n) :
+    reciprocalFactors n = n % 4 * n :: reciprocalFactors (n / 4 + 1) := by
+  have h0 : n ≠ 0
+  · rintro rfl
+    norm_num at h2
+  conv_lhs => unfold reciprocalFactors
+  rw [dif_neg h0, dif_neg h1, if_neg (Nat.odd_iff_not_even.mp h2)]
```

Comment:
This `unusedHavesSuffices` weirdness is probably worth filing a bug somewhere about, or asking on Zulip.